functions/database/database.js

```diff
@@ -0,0 +1,414 @@
+ *
+ * @param {string} releaseId The ID of the release to fetch.
+ * @throws {Error} If there is no release with the given ID.
+ * @return {Promise<Object>} A promise that resolves to the release data.
+ */
+async function getReleaseData(releaseId) {
+  const releaseSnapshot = await db.collection("releases")
+      .doc(releaseId)
+      .get();
+
+  if (!releaseSnapshot.exists) {
+    throw new Error(`No release found with ID: ${releaseId}`);
+  }
+
+  const releaseData = releaseSnapshot.data();
+
+  return releaseData;
+}
+
+/**
+ * Prepare a single release object for storage into Firestore.
+ *
+ * @param {Object} release - A release object
+ * @return {Object} - A release data object ready for storage
+ */
+function releaseToFirestoreObject(release) {
+  const snapshotBranchName = `releases/${release.releaseName}`;
+  const releaseBranchName = `releases/${release.releaseName}.release`;
+  const snapshotBranchLink = `${REPO_URL}/tree/${snapshotBranchName}`;
+  const releaseBranchLink = `${REPO_URL}/tree/${releaseBranchName}`;
+
+  return {
+    state: RELEASE_STATES.SCHEDULED,
+    releaseName: release.releaseName,
+    releaseOperator: release.releaseOperator,
+    codeFreezeDate: release.codeFreezeDate,
+    releaseDate: release.releaseDate,
+    snapshotBranchName: snapshotBranchName,
+    snapshotBranchLink: snapshotBranchLink,
+    releaseBranchName: releaseBranchName,
+    releaseBranchLink: releaseBranchLink,
+    isComplete: false,
+    buildArtifactStatus: "",
+    buildArtifactConclusion: "",
+    buildArtifactLink: "",
+    buildArtifactJobId: "",
+  };
+}
+
+/**
+ * Set a release object into Firestore batch for further processing.
+ *
+ *
+ * Note that this will not commit the change, it merely adds it to
+ * the given batch.
+ *
+ * @param {admin.firestore.WriteBatch} batch - The Firestore batch instance
+ * @param {Object} release - A release object
+ */
+function batchSetRelease(batch, release) {
+  const newReleaseRef = db.collection("releases").doc();
+  const releaseData = releaseToFirestoreObject(release);
+  batch.set(newReleaseRef, releaseData);
+}
+
+/**
+ * Write new releases to Firestore.
+ *
+ * Assumes that the contents of the releases are validated. newReleases is
+ * an array of release objects that have the following structure:
+ * {
+ *   releaseName: string,
+ *   releaseOperator: string,
+ *   codeFreezeDate: Date,
+ *   releaseDate: Date,
+ * }
+ *
+ * @param {Object} newReleases - Releases to store in Firestore
+ */
+async function setReleases(newReleases) {
+  validateNewReleasesStructure(newReleases);
+
+  const batch = db.batch();
+
+  newReleases.forEach((release) => batchSetRelease(batch, release));
+
+  await batch.commit();
+}
+/**
+ * Delete the releases in Firestore with code freeze dates that are in the
+ * future.
+ *
+ * This is intented to be used when scheduling new upcoming releases,
+ * but we need to delete the old upcoming releases before.
+ */
+async function deleteUpcomingReleases() {
```

Comment:
Is this still planned to be used by the frontend? It seems dangerous in cases where client state  has deviated from  DB state

Reply:
No, this will no longer be used by the frontend- it's definitely no good.

I just noticed it's not called anywhere anymore, it's just dead code- I'll remove it.

Reply:
Removed in https://github.com/firebase/firebase-release-dashboard/pull/8/commits/fad927bc0b9bd4bc52eeb3ac14147fb3bc90ca6f